### PR TITLE
Add Improved AutoEncoder model.

### PR DESCRIPTION
**Pull request by acsaga**
Add a new autoencoder model, I call it Improved AutoEncoder (IAE).
In the same training time, this model generates better(sharper and natural-looking)  faces than the original model.

In IAE, autoencoder_A and autoencoder_B use the same encoder and decoder.
Between the encoder and decoder, IAE explicitly adds intermidiate layers for latent varibles.

An (ugly) illustration for the model structure:
image

### CODE REVIEW

**Comment by Clorr** [comment]
Hi @acsaga 
Thank you for this awesome PR, it is always a pleasure to see a new model emerge!
I'm going to test it ASAP. Just a remark, we are moving to a new folder structure in the latest commits. You can check [my repo](https://github.com/Clorr/faceswap/blob/master/plugins/Model_Original/) to see how it is now. Recommended structure would be for you:
- Create `Model_IAE` folder in plugins
- Create __init__.py like [this one](https://github.com/Clorr/faceswap/blob/master/plugins/Model_Original/__init__.py)
- Put your files in this folder

Let me know if you need help or if you want me to do this for you

**Comment by Clorr** [comment]
Thanks for doing the changes. however I think you forgot to add the files. The `.gitignore` is buggy so you have to add the new files manually. I don't know which tool you are using, but with git cli, you have to do `git add -f *.py` from the folder where your files are ;-)

**Comment by acsaga** [comment]
Thanks for your tips :)
Now it should be fine.

**Comment by Clorr** [comment]
Seems perfect to me ;-) I can merge this right now or do you want to wait for some feedback? (If I merge, I suggest you add an issue so that others are aware of this new plugin)

**Comment by Apollo122** [comment]
An example output, maybe side by side comparison would have been nice.

**Comment by iperov** [comment]
I launched train with Daddario. Will see difference.

**Comment by acsaga** [comment]
@Apollo122 FYI, here's a quick comparison:

Face A: [aino kishi](https://www.google.com/search?q=Aino+Kishi&source=lnms&tbm=isch&sa=X&ved=0ahUKEwjek7nQn93ZAhXnzlQKHTD8AdIQ_AUICigB&biw=2560&bih=1334)

Face B: [gakki](https://www.google.com/search?q=gakki&source=lnms&tbm=isch&sa=X&ved=0ahUKEwjgv9Szn93ZAhXQ31QKHfZ6DsQQ_AUICigB&biw=2560&bih=1334)



**Original Model：**
loss：
faceswap-original-loss
output：
faceswap-original-image

**IAE Model：**
loss：
faceswap-iae-loss
output：
faceswap-iae-image

I trained the IAE model less time than the original model, and the original model has smaller loss function value. But you can see that the output of IAE model has better quality.

**Comment by acsaga** [comment]
@Clorr Feedback is welcome. 
I will try this model on the Trump/Cage dataset this weekend and post the result.

**Comment by Apollo122** [comment]
@acsaga thanks for the comparison.
Does IAE need a separate converter? Can we use the Masked converter to merge ?

**Comment by ruah1984** [comment]
Can the existing model train with this IE model again ?

**Comment by kellurian** [comment]
You can add the multiGpu model calls in your model files as well and they work fine, so when we get it all interposed into one model file it should work great. bryanlyon seems to be doing this well in #256 . I dowloaded and tested your model with my two GPU's and the multiGPU additons and holy crap do they run fast! Funny enough, the old model ran my GTX 980 TI's at about 40% but it was up and down, this model runs it at about 70% but it is a seems to be much more constant. But they don't work with a previously tested model @ruah1984, it'll start the training over. I like this new model idea!

**Comment by Clorr** [comment]
@acsaga for my knowledge, is there some theoritical background for this? or is it the result of your experiment? It is a kind of variational autoencoder, isn't it?

**Comment by iperov** [comment]
AMAZING RESULT

**Comment by iperov** [comment]
after 18 hours of training there is no more loss changes (every 10 min):



look at comparison video SFW
convert options same
https://webm.video/i/sy8zXG.mp4


at left - Original but my modified "hi-res" decoder:
```
def Decoder(self):
        input_ = Input(shape=(8, 8, 512))
        x = input_
        x = self.upscale(512)(x)
        x = self.upscale(256)(x)
        x = self.upscale(128)(x)
        x = self.upscale(64)(x)
        x = self.upscale(32)(x)
        x = Conv2D(3, kernel_size=5, padding='same', activation='sigmoid')(x)
        x = NearestNeighborDownsampler()(x)
        x = BicubicDownsampler()(x)        
        return KerasModel(input_, x)
```
so I got more detailed eyes

at right - IAE but decoder same as Original.

now compare problem places:
vegas130_2018-03-09_14-45-11
vegas130_2018-03-09_14-42-13
vegas130_2018-03-09_14-48-03

IAE better.

Now I will launch IAE but with "hi-res" decoder.

**Comment by acsaga** [comment]
@Apollo122 Yes, this model use the same converter as the Original model.
@kellurian Thanks, I will add the multiGPU option and test it later.

@Clorr Yes, theoretically it can be interpreted as some kind of variational autoencoder. 

The intuitive idea behind this model is that: A face has two types of information - appearance and facial expression. 
What IAE model does is:
1. Extract face information from a face image. (Encoder)
2. Split the face information to two parts: facial expression and appearance. (intermediate layers, inter_A for appearance of face A, inter_B for appearance of face B, inter_both for facial expression of A and B)
3. Combine face_A's facial expression and face_B's appearance as new information. (Concatenate intermediate layers)
4. Generate new face image from new information. (Decoder)

Actually the original model does the same thing. IAE Model gets better result because it separates the steps into different layers. So that every layer focuses on only one task.
The original model mixes the steps: encoder does task No.1 and stores facial expressions. Decoder stores appearance and does task No.4. So it's hard to train.

**Comment by acsaga** [comment]
@iperov Glad to hear it works 😄

**Comment by Clorr** [comment]
@acsaga thanks for this information. 

I'm always fascinated to see how versatile NN are... How many faces do you think we could store in one model? Because in this case, we could add more intermediate layers for more people, don't you think?

Btw, if you have some inspiration, feel free to give more information in your __init.py__ about yourself, credits, license and so on...

**Comment by ruah1984** [comment]
@acsaga ,if each layer of encoder we manage different information , when all information merge and generate (decoder)new face, will have more lost?

**Comment by ruah1984** [comment]
I try this yesterday night, is fast to get the clear and clean image compare to original model. Don't know if retrain decoder + IAE can be work together or not

**Comment by bryanlyon** [comment]
This is still a good add.  I've modified it to support the new plugin loader, and the gan v2.0 update at https://github.com/bryanlyon/faceswap/tree/multi_gpu_iae .  I will create a pull request if desired, but it requires #272 to be merged first.

**Comment by iperov** [comment]
got much different result vs original in man to man conversion

source man:


dest man:
out00042


Original:
out00047

IAE:
out00047 2

IAE constructs absolutely new unrecognizable man. Also it adds some mustach.

**Comment by acsaga** [comment]
18 hours training (1080Ti) on cage/trump task:

image

image

Please merge this pull request when it's ok for you. I think it's time to let more people use and test this model ;)

@Clorr @ruah1984 Principally, if there are enough different faces for training, I think the encoder in iae can be an universal face encoder (extract facial expression and appearance from any face), and the decoder in iae can be an universal face decoder (reconstruct face image from arbitrary combination of facial expression and appearance). Then Yes we can train many intermediate layers for many people, and swap faces for any two of them 🤔 

@bryanlyon Thanks for adding multi gpu support for the model 👍 

@iperov I guess the difference is because the model need more faces of "source man" (with different facial expression) to help the model splits latent variables into "facial expression" and "appearance“.
In your example, the unrecognizable man looks like a mix of source man and dest man, because the model failed to distinguish "facial expression" and "appearance“ information, and categorized too much information to "facial expression" layer (for example, the mustache).

**Comment by iperov** [comment]
@acsaga source man 3000 photos.

**Comment by Clorr** [comment]
Important notice: I renamed the model files so that users who try this don't overwrite their original model

Reanme your existing IAE model files if you did train some models previously! (I added a "IAE_" prefix)

**Comment by Kirin-kun** [comment]
@iperov the eyes you produced with your "hi-res" decoder are awesome. Any chance to add it to the repo?

**Comment by PhenomenalOnee** [comment]
Where i can find the pretrained IAE model

**Comment by torzdf** [comment]
There are no pretrained models

**Comment by PhenomenalOnee** [comment]
How can i train the IAE model can you @torzdf, provide me any scripts for data preparation and training. And also how is the predicted face and mask is merged on the destination face

**Comment by PhenomenalOnee** [comment]
Can someone explain how to train the model IAE step by step
currently i am not able to interpretate  how faceswap.py extract works which functions it calls etc

**Comment by torzdf** [comment]
These are better places for this kind of info:

https://github.com/deepfakes/faceswap/blob/master/USAGE.md

https://faceswap.dev

**Comment by PhenomenalOnee** [comment]
Thanks @torzdf Can you explain how the predicted face is wrapped and swapped with the source face and how mask is formed

**Comment by kvrooman** [comment]
These sorts of questions are best asked on the support forum or on the discord server. GitHub is typically used for bug reporting and testing features.

**Comment by AliDjango** [comment]
> 
> 
> @Clorr Feedback is welcome.
> I will try this model on the Trump/Cage dataset this weekend and post the result.

Is this dataset available to download?I want a dataset, so I can see how well different models train by time before making my own training

**Comment by torzdf** [comment]
Faceswap does not provide pre-trained models.

**Comment by GffdesgjbdsXc** [comment]
Hi
